include/aws/common/array_list_with_map.h

```diff
@@ -0,0 +1,82 @@
+#ifndef AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+#define AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/common/common.h>
+#include <aws/common/hash_table.h>
+
+struct aws_array_list_with_map {
+    struct aws_array_list list;
+    struct aws_hash_table map; /* map from the element to the index in the array */
+};
+
+AWS_EXTERN_C_BEGIN
+
+/* TODO: CBMC verification. */
+/**
+ * Initialize the array_list with map structure, which support constant time of insert, remove and get random element
+ * from the data structure.
+ *
+ * The underlying hash map will use hash_fn to compute the hash of each element. equals_fn to compute equality of two
+ * keys.
+ *
+ * @param list_with_map             Pointer of structure to initialize with
+ * @param allocator                 Allocator
+ * @param hash_fn                   Compute the hash of each element
+ * @param equals_fn                 Compute equality of two keys
+ * @param destroy_element_fn        Called when the element is removed
+ * @param item_size                 Size of each element
+ * @param initial_item_allocation   The initial number of item to allocate.
+ * @return AWS_OP_ERR if any fails to initialize, AWS_OP_SUCCESS on success.
+ */
+AWS_COMMON_API
+int aws_array_list_with_map_init(
+    struct aws_array_list_with_map *list_with_map,
+    struct aws_allocator *allocator,
+    aws_hash_fn *hash_fn,
+    aws_hash_callback_eq_fn *equals_fn,
+    aws_hash_callback_destroy_fn *destroy_element_fn,
+    size_t item_size,
+    size_t initial_item_allocation);
+
+AWS_COMMON_API
+void aws_array_list_with_map_clean_up(struct aws_array_list_with_map *list_with_map);
+
+/**
+ * Insert the element to the end of the array list. A map from the element to the index of it to the hash table.
```

Comment:
adds an element to the set, no need to mention array list. What does this function do if an element already exists? Do you want an optional was-created parameter?

https://github.com/awslabs/aws-c-common/blob/d9031cf4f02fe4d8c24a53bb6c8df67ce8be9f0c/include/aws/common/hash_table.h#L269-L270

Reply:
Well, we do tell the implementation to the user. Like the when we initial the data structure, the hash functions are details but user needs to know about. 
I have no strong feeling about the was created, but I thought about to support duplicates in the future if we need. So, a was created will be weird in that case.